In [7]:
import pandas as pd

In [5]:
from google.colab import files

uploaded = files.upload()

Saving SampleSuperstore.csv to SampleSuperstore (1).csv


In [8]:
df = pd.read_csv("SampleSuperstore.csv")

In [7]:
df

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
0,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Bookcases,261.9600,2,0.00,41.9136
1,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Chairs,731.9400,3,0.00,219.5820
2,Second Class,Corporate,United States,Los Angeles,California,90036,West,Office Supplies,Labels,14.6200,2,0.00,6.8714
3,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Furniture,Tables,957.5775,5,0.45,-383.0310
4,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Office Supplies,Storage,22.3680,2,0.20,2.5164
...,...,...,...,...,...,...,...,...,...,...,...,...,...
9989,Second Class,Consumer,United States,Miami,Florida,33180,South,Furniture,Furnishings,25.2480,3,0.20,4.1028
9990,Standard Class,Consumer,United States,Costa Mesa,California,92627,West,Furniture,Furnishings,91.9600,2,0.00,15.6332
9991,Standard Class,Consumer,United States,Costa Mesa,California,92627,West,Technology,Phones,258.5760,2,0.20,19.3932
9992,Standard Class,Consumer,United States,Costa Mesa,California,92627,West,Office Supplies,Paper,29.6000,4,0.00,13.3200


In [9]:
df.head()

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
0,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Bookcases,261.9600,2,0.00,41.9136
1,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Chairs,731.9400,3,0.00,219.5820
2,Second Class,Corporate,United States,Los Angeles,California,90036,West,Office Supplies,Labels,14.6200,2,0.00,6.8714
3,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Furniture,Tables,957.5775,5,0.45,-383.0310
4,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Office Supplies,Storage,22.3680,2,0.20,2.5164


In [10]:
df.shape

(9994, 13)

In [11]:
df.columns

Index(['Ship Mode', 'Segment', 'Country', 'City', 'State', 'Postal Code',
       'Region', 'Category', 'Sub-Category', 'Sales', 'Quantity', 'Discount',
       'Profit'],
      dtype='object')

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Ship Mode     9994 non-null   object 
 1   Segment       9994 non-null   object 
 2   Country       9994 non-null   object 
 3   City          9994 non-null   object 
 4   State         9994 non-null   object 
 5   Postal Code   9994 non-null   int64  
 6   Region        9994 non-null   object 
 7   Category      9994 non-null   object 
 8   Sub-Category  9994 non-null   object 
 9   Sales         9994 non-null   float64
 10  Quantity      9994 non-null   int64  
 11  Discount      9994 non-null   float64
 12  Profit        9994 non-null   float64
dtypes: float64(3), int64(2), object(8)
memory usage: 1015.1+ KB


In [13]:
df.isnull().sum()

,0
Ship Mode,0
Segment,0
Country,0
City,0
State,0
Postal Code,0
Region,0
Category,0
Sub-Category,0
Sales,0


In [17]:
df.isnull().sum().sum()

np.int64(0)

In [18]:
df.isnull().sum().sum()

np.int64(0)

In [19]:
df[df.duplicated(keep=False)]

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
568,Standard Class,Corporate,United States,Seattle,Washington,98105,West,Office Supplies,Paper,19.440,3,0.0,9.3312
591,Standard Class,Consumer,United States,Salem,Oregon,97301,West,Office Supplies,Paper,10.368,2,0.2,3.6288
935,Standard Class,Home Office,United States,Philadelphia,Pennsylvania,19120,East,Office Supplies,Paper,15.552,3,0.2,5.4432
950,Standard Class,Home Office,United States,Philadelphia,Pennsylvania,19120,East,Office Supplies,Paper,15.552,3,0.2,5.4432
1186,Standard Class,Corporate,United States,Seattle,Washington,98103,West,Office Supplies,Paper,25.920,4,0.0,12.4416
1479,Standard Class,Consumer,United States,San Francisco,California,94122,West,Office Supplies,Paper,25.920,4,0.0,12.4416
2803,Standard Class,Consumer,United States,San Francisco,California,94122,West,Office Supplies,Paper,12.840,3,0.0,5.7780
2807,Second Class,Consumer,United States,Seattle,Washington,98115,West,Office Supplies,Paper,12.960,2,0.0,6.2208
2836,Standard Class,Consumer,United States,Los Angeles,California,90036,West,Office Supplies,Paper,19.440,3,0.0,9.3312
3127,Standard Class,Consumer,United States,New York City,New York,10011,East,Office Supplies,Paper,49.120,4,0.0,23.0864


In [20]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 17


In [21]:
df[df.duplicated(keep=False)]

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
568,Standard Class,Corporate,United States,Seattle,Washington,98105,West,Office Supplies,Paper,19.440,3,0.0,9.3312
591,Standard Class,Consumer,United States,Salem,Oregon,97301,West,Office Supplies,Paper,10.368,2,0.2,3.6288
935,Standard Class,Home Office,United States,Philadelphia,Pennsylvania,19120,East,Office Supplies,Paper,15.552,3,0.2,5.4432
950,Standard Class,Home Office,United States,Philadelphia,Pennsylvania,19120,East,Office Supplies,Paper,15.552,3,0.2,5.4432
1186,Standard Class,Corporate,United States,Seattle,Washington,98103,West,Office Supplies,Paper,25.920,4,0.0,12.4416
1479,Standard Class,Consumer,United States,San Francisco,California,94122,West,Office Supplies,Paper,25.920,4,0.0,12.4416
2803,Standard Class,Consumer,United States,San Francisco,California,94122,West,Office Supplies,Paper,12.840,3,0.0,5.7780
2807,Second Class,Consumer,United States,Seattle,Washington,98115,West,Office Supplies,Paper,12.960,2,0.0,6.2208
2836,Standard Class,Consumer,United States,Los Angeles,California,90036,West,Office Supplies,Paper,19.440,3,0.0,9.3312
3127,Standard Class,Consumer,United States,New York City,New York,10011,East,Office Supplies,Paper,49.120,4,0.0,23.0864


In [22]:
df[df.duplicated(keep=False)].shape

(34, 13)

In [23]:
df_clean = df.copy()

In [24]:
df_clean = df_clean.drop_duplicates()

In [25]:
print("Duplicate rows after cleaning:", df_clean.duplicated().sum())

Duplicate rows after cleaning: 0


In [26]:
print("Original rows:", df.shape[0])
print("Cleaned rows:", df_clean.shape[0])
print("Rows removed:", df.shape[0] - df_clean.shape[0])

Original rows: 9994
Cleaned rows: 9977
Rows removed: 17


In [27]:
text_columns = df_clean.select_dtypes(include="object").columns

print(text_columns)

Index(['Ship Mode', 'Segment', 'Country', 'City', 'State', 'Region',
       'Category', 'Sub-Category'],
      dtype='object')


In [28]:
for col in text_columns:
    spaces = (df_clean[col] != df_clean[col].str.strip()).sum()
    print(col, ":", spaces)

Ship Mode : 0
Segment : 0
Country : 0
City : 0
State : 0
Region : 0
Category : 0
Sub-Category : 0


In [29]:
for col in text_columns:
    print("\n---", col, "---")
    print(df_clean[col].value_counts())


--- Ship Mode ---
Ship Mode
Standard Class    5955
Second Class      1943
First Class       1537
Same Day           542
Name: count, dtype: int64

--- Segment ---
Segment
Consumer       5183
Corporate      3015
Home Office    1779
Name: count, dtype: int64

--- Country ---
Country
United States    9977
Name: count, dtype: int64

--- City ---
City
New York City    914
Los Angeles      746
Philadelphia     536
San Francisco    506
Seattle          424
                ... 
Abilene            1
Montebello         1
Kissimmee          1
Danbury            1
Springdale         1
Name: count, Length: 531, dtype: int64

--- State ---
State
California              1996
New York                1127
Texas                    983
Pennsylvania             586
Washington               502
Illinois                 491
Ohio                     468
Florida                  383
Michigan                 254
North Carolina           249
Arizona                  224
Virginia                 224
Georgia    

In [30]:
df_clean[df_clean["Ship Mode"] == "Ship Mode"]

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit


In [31]:
for col in text_columns:
    print("\n---", col, "---")
    print(df_clean[col].value_counts().head(10))


--- Ship Mode ---
Ship Mode
Standard Class    5955
Second Class      1943
First Class       1537
Same Day           542
Name: count, dtype: int64

--- Segment ---
Segment
Consumer       5183
Corporate      3015
Home Office    1779
Name: count, dtype: int64

--- Country ---
Country
United States    9977
Name: count, dtype: int64

--- City ---
City
New York City    914
Los Angeles      746
Philadelphia     536
San Francisco    506
Seattle          424
Houston          375
Chicago          313
Columbus         221
San Diego        170
Springfield      163
Name: count, dtype: int64

--- State ---
State
California        1996
New York          1127
Texas              983
Pennsylvania       586
Washington         502
Illinois           491
Ohio               468
Florida            383
Michigan           254
North Carolina     249
Name: count, dtype: int64

--- Region ---
Region
West       3193
East       2845
Central    2319
South      1620
Name: count, dtype: int64

--- Category ---
Catego

In [32]:
df_clean[df_clean["Ship Mode"] == "Ship Mode"]

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit


In [33]:
df_clean[df_clean["Ship Mode"] == "Ship Mode"].T

""
Ship Mode
Segment
Country
City
State
Postal Code
Region
Category
Sub-Category
Sales


In [34]:
header_row = df_clean[df_clean["Ship Mode"] == "Ship Mode"]

print(header_row.to_dict("records"))

[]


In [35]:
print("Repeated header rows:",
      (df_clean["Ship Mode"] == "Ship Mode").sum())

Repeated header rows: 0


In [36]:
df_clean["Ship Mode"].value_counts()

,count
Ship Mode,
Standard Class,5955
Second Class,1943
First Class,1537
Same Day,542


In [37]:
df_clean["Ship Mode"].unique()

array(['Second Class', 'Standard Class', 'First Class', 'Same Day'],
      dtype=object)

In [38]:
header_mask = df_clean.eq(df_clean.columns).all(axis=1)

print("Repeated header rows:", header_mask.sum())

Repeated header rows: 0


In [39]:
for col in text_columns:
    print("\n---", col, "---")
    print(df_clean[col].value_counts())


--- Ship Mode ---
Ship Mode
Standard Class    5955
Second Class      1943
First Class       1537
Same Day           542
Name: count, dtype: int64

--- Segment ---
Segment
Consumer       5183
Corporate      3015
Home Office    1779
Name: count, dtype: int64

--- Country ---
Country
United States    9977
Name: count, dtype: int64

--- City ---
City
New York City    914
Los Angeles      746
Philadelphia     536
San Francisco    506
Seattle          424
                ... 
Abilene            1
Montebello         1
Kissimmee          1
Danbury            1
Springdale         1
Name: count, Length: 531, dtype: int64

--- State ---
State
California              1996
New York                1127
Texas                    983
Pennsylvania             586
Washington               502
Illinois                 491
Ohio                     468
Florida                  383
Michigan                 254
North Carolina           249
Arizona                  224
Virginia                 224
Georgia    

In [40]:
df_clean.dtypes

,0
Ship Mode,object
Segment,object
Country,object
City,object
State,object
Postal Code,int64
Region,object
Category,object
Sub-Category,object
Sales,float64


In [41]:
df["Sales"].mean()

np.float64(229.85800083049833)

In [42]:
df["Sales"].sum()

np.float64(2297200.8603000003)

In [43]:
df_clean.dtypes

,0
Ship Mode,object
Segment,object
Country,object
City,object
State,object
Postal Code,int64
Region,object
Category,object
Sub-Category,object
Sales,float64


In [44]:
df_clean.describe()

,Postal Code,Sales,Quantity,Discount,Profit
count,9977.000000,9977.000000,9977.000000,9977.000000,9977.00000
mean,55154.964117,230.148902,3.790719,0.156278,28.69013
std,32058.266816,623.721409,2.226657,0.206455,234.45784
min,1040.000000,0.444000,1.000000,0.000000,-6599.97800
25%,23223.000000,17.300000,2.000000,0.000000,1.72620
50%,55901.000000,54.816000,3.000000,0.200000,8.67100
75%,90008.000000,209.970000,5.000000,0.200000,29.37200
max,99301.000000,22638.480000,14.000000,0.800000,8399.97600


In [45]:
print("Minimum Quantity:", df_clean["Quantity"].min())
print("Maximum Quantity:", df_clean["Quantity"].max())

Minimum Quantity: 1
Maximum Quantity: 14


In [46]:
print("Minimum Discount:", df_clean["Discount"].min())
print("Maximum Discount:", df_clean["Discount"].max())

Minimum Discount: 0.0
Maximum Discount: 0.8


In [47]:
print("Minimum Sales:", df_clean["Sales"].min())
print("Maximum Sales:", df_clean["Sales"].max())

Minimum Sales: 0.444
Maximum Sales: 22638.48


In [48]:
print("Minimum Quantity:", df_clean["Quantity"].min())
print("Maximum Quantity:", df_clean["Quantity"].max())

Minimum Quantity: 1
Maximum Quantity: 14


In [49]:
print("Zero Quantity:", (df_clean["Quantity"] == 0).sum())
print("Negative Quantity:", (df_clean["Quantity"] < 0).sum())

Zero Quantity: 0
Negative Quantity: 0


In [50]:
print("Minimum Discount:", df_clean["Discount"].min())
print("Maximum Discount:", df_clean["Discount"].max())

Minimum Discount: 0.0
Maximum Discount: 0.8


In [51]:
print("Discount below 0:", (df_clean["Discount"] < 0).sum())
print("Discount above 1:", (df_clean["Discount"] > 1).sum())

Discount below 0: 0
Discount above 1: 0


In [52]:
print("Minimum Profit:", df_clean["Profit"].min())
print("Maximum Profit:", df_clean["Profit"].max())

Minimum Profit: -6599.978
Maximum Profit: 8399.976


In [53]:
print("Negative Profit:", (df_clean["Profit"] < 0).sum())

Negative Profit: 1869


In [54]:
print("Zero Profit:", (df_clean["Profit"] == 0).sum())

Zero Profit: 65


In [55]:
print("Minimum Profit:", df_clean["Profit"].min())
print("Maximum Profit:", df_clean["Profit"].max())

Minimum Profit: -6599.978
Maximum Profit: 8399.976


In [56]:
print("Negative Profit:", (df_clean["Profit"] < 0).sum())

Negative Profit: 1869


In [57]:
print("Minimum Profit:", df_clean["Profit"].min())
print("Maximum Profit:", df_clean["Profit"].max())

Minimum Profit: -6599.978
Maximum Profit: 8399.976


In [58]:
df_clean.describe()

,Postal Code,Sales,Quantity,Discount,Profit
count,9977.000000,9977.000000,9977.000000,9977.000000,9977.00000
mean,55154.964117,230.148902,3.790719,0.156278,28.69013
std,32058.266816,623.721409,2.226657,0.206455,234.45784
min,1040.000000,0.444000,1.000000,0.000000,-6599.97800
25%,23223.000000,17.300000,2.000000,0.000000,1.72620
50%,55901.000000,54.816000,3.000000,0.200000,8.67100
75%,90008.000000,209.970000,5.000000,0.200000,29.37200
max,99301.000000,22638.480000,14.000000,0.800000,8399.97600


In [59]:
Q1 = df_clean["Sales"].quantile(0.25)
Q3 = df_clean["Sales"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)

Q1: 17.3
Q3: 209.97
IQR: 192.67
Lower Bound: -271.705
Upper Bound: 498.975


In [60]:
sales_outliers = df_clean[
    (df_clean["Sales"] < lower_bound) |
    (df_clean["Sales"] > upper_bound)
]

print("Potential Sales outliers:", len(sales_outliers))


Potential Sales outliers: 1167


In [61]:
sales_outliers[["Sales", "Quantity", "Discount", "Profit"]].head(20)

,Sales,Quantity,Discount,Profit
1,731.9400,3,0.00,219.5820
3,957.5775,5,0.45,-383.0310
7,907.1520,6,0.20,90.7152
10,1706.1840,9,0.20,85.3092
11,911.4240,4,0.20,68.3568
16,665.8800,6,0.00,13.3176
24,1044.6300,3,0.00,240.2649
27,3083.4300,7,0.50,-1665.0522
35,1097.5440,7,0.20,123.4737
38,532.3992,3,0.32,-46.9764


In [62]:
df_clean.nlargest(10, "Sales")[[
    "Sales",
    "Quantity",
    "Discount",
    "Profit"
]]

,Sales,Quantity,Discount,Profit
2697,22638.480,6,0.5,-1811.0784
6826,17499.950,5,0.0,8399.9760
8153,13999.960,4,0.0,6719.9808
2623,11199.968,4,0.2,3919.9888
4190,10499.970,3,0.0,5039.9856
9039,9892.740,13,0.0,4946.3700
4098,9449.950,5,0.0,4630.4755
4277,9099.930,7,0.0,2365.9818
8488,8749.950,5,0.0,2799.9840
6425,8399.976,4,0.4,1119.9968


In [63]:
df_clean.nsmallest(10, "Sales")[[
    "Sales",
    "Quantity",
    "Discount",
    "Profit"
]]

,Sales,Quantity,Discount,Profit
4101,0.444,1,0.8,-1.1100
9292,0.556,1,0.8,-0.9452
8658,0.836,1,0.8,-1.3376
4711,0.852,1,0.7,-0.5964
2106,0.876,1,0.8,-1.4016
7548,0.898,1,0.8,-1.5715
8033,0.984,2,0.8,-1.4760
2761,0.990,1,0.0,0.4356
8024,1.044,1,0.8,-1.8270
976,1.080,2,0.7,-0.7920


In [64]:
Q1_profit = df_clean["Profit"].quantile(0.25)
Q3_profit = df_clean["Profit"].quantile(0.75)

IQR_profit = Q3_profit - Q1_profit

lower_profit = Q1_profit - 1.5 * IQR_profit
upper_profit = Q3_profit + 1.5 * IQR_profit

print("Q1:", Q1_profit)
print("Q3:", Q3_profit)
print("IQR:", IQR_profit)
print("Lower Bound:", lower_profit)
print("Upper Bound:", upper_profit)

Q1: 1.7262
Q3: 29.372
IQR: 27.6458
Lower Bound: -39.7425
Upper Bound: 70.8407


In [65]:
profit_outliers = df_clean[
    (df_clean["Profit"] < lower_profit) |
    (df_clean["Profit"] > upper_profit)
]

print("Potential Profit outliers:", len(profit_outliers))

Potential Profit outliers: 1881


In [66]:
profit_outliers[[
    "Sales",
    "Quantity",
    "Discount",
    "Profit"
]].head(20)

,Sales,Quantity,Discount,Profit
1,731.9400,3,0.00,219.5820
3,957.5775,5,0.45,-383.0310
7,907.1520,6,0.20,90.7152
10,1706.1840,9,0.20,85.3092
13,407.9760,3,0.20,132.5922
14,68.8100,5,0.80,-123.8580
24,1044.6300,3,0.00,240.2649
27,3083.4300,7,0.50,-1665.0522
35,1097.5440,7,0.20,123.4737
36,190.9200,5,0.60,-147.9630


In [67]:
df_clean.nlargest(10, "Sales")[[
    "Sales", "Quantity", "Discount", "Profit"
]]

,Sales,Quantity,Discount,Profit
2697,22638.480,6,0.5,-1811.0784
6826,17499.950,5,0.0,8399.9760
8153,13999.960,4,0.0,6719.9808
2623,11199.968,4,0.2,3919.9888
4190,10499.970,3,0.0,5039.9856
9039,9892.740,13,0.0,4946.3700
4098,9449.950,5,0.0,4630.4755
4277,9099.930,7,0.0,2365.9818
8488,8749.950,5,0.0,2799.9840
6425,8399.976,4,0.4,1119.9968


In [68]:
Q1_profit = df_clean["Profit"].quantile(0.25)
Q3_profit = df_clean["Profit"].quantile(0.75)

IQR_profit = Q3_profit - Q1_profit

lower_profit = Q1_profit - 1.5 * IQR_profit
upper_profit = Q3_profit + 1.5 * IQR_profit

print("Q1:", Q1_profit)
print("Q3:", Q3_profit)
print("IQR:", IQR_profit)
print("Lower Bound:", lower_profit)
print("Upper Bound:", upper_profit)

Q1: 1.7262
Q3: 29.372
IQR: 27.6458
Lower Bound: -39.7425
Upper Bound: 70.8407


In [69]:
profit_outliers = df_clean[
    (df_clean["Profit"] < lower_profit) |
    (df_clean["Profit"] > upper_profit)
]

print("Potential Profit outliers:", len(profit_outliers))

Potential Profit outliers: 1881


In [70]:
profit_outliers[[
    "Sales",
    "Quantity",
    "Discount",
    "Profit"
]].head(20)

,Sales,Quantity,Discount,Profit
1,731.9400,3,0.00,219.5820
3,957.5775,5,0.45,-383.0310
7,907.1520,6,0.20,90.7152
10,1706.1840,9,0.20,85.3092
13,407.9760,3,0.20,132.5922
14,68.8100,5,0.80,-123.8580
24,1044.6300,3,0.00,240.2649
27,3083.4300,7,0.50,-1665.0522
35,1097.5440,7,0.20,123.4737
36,190.9200,5,0.60,-147.9630


In [71]:
df_clean.nlargest(10, "Profit")[[
    "Sales",
    "Quantity",
    "Discount",
    "Profit"
]]

,Sales,Quantity,Discount,Profit
6826,17499.950,5,0.0,8399.9760
8153,13999.960,4,0.0,6719.9808
4190,10499.970,3,0.0,5039.9856
9039,9892.740,13,0.0,4946.3700
4098,9449.950,5,0.0,4630.4755
2623,11199.968,4,0.2,3919.9888
509,6354.950,5,0.0,3177.4750
8488,8749.950,5,0.0,2799.9840
7666,5399.910,9,0.0,2591.9568
6520,5443.960,4,0.0,2504.2216


In [72]:
df_clean.nsmallest(10, "Profit")[[
    "Sales",
    "Quantity",
    "Discount",
    "Profit"
]]

,Sales,Quantity,Discount,Profit
7772,4499.985,5,0.7,-6599.9780
683,7999.980,4,0.5,-3839.9904
9774,2177.584,8,0.8,-3701.8928
3011,2549.985,5,0.7,-3399.9800
4991,1889.990,5,0.8,-2929.4845
3151,1799.994,2,0.7,-2639.9912
5310,1525.188,6,0.8,-2287.7820
9639,4297.644,13,0.4,-1862.3124
1199,1088.792,4,0.8,-1850.9464
2697,22638.480,6,0.5,-1811.0784


In [73]:
df_clean.nlargest(10, "Profit")[[
    "Sales", "Quantity", "Discount", "Profit"
]]

,Sales,Quantity,Discount,Profit
6826,17499.950,5,0.0,8399.9760
8153,13999.960,4,0.0,6719.9808
4190,10499.970,3,0.0,5039.9856
9039,9892.740,13,0.0,4946.3700
4098,9449.950,5,0.0,4630.4755
2623,11199.968,4,0.2,3919.9888
509,6354.950,5,0.0,3177.4750
8488,8749.950,5,0.0,2799.9840
7666,5399.910,9,0.0,2591.9568
6520,5443.960,4,0.0,2504.2216


In [74]:
df_clean.nsmallest(10, "Profit")[[
    "Sales", "Quantity", "Discount", "Profit"
]]

,Sales,Quantity,Discount,Profit
7772,4499.985,5,0.7,-6599.9780
683,7999.980,4,0.5,-3839.9904
9774,2177.584,8,0.8,-3701.8928
3011,2549.985,5,0.7,-3399.9800
4991,1889.990,5,0.8,-2929.4845
3151,1799.994,2,0.7,-2639.9912
5310,1525.188,6,0.8,-2287.7820
9639,4297.644,13,0.4,-1862.3124
1199,1088.792,4,0.8,-1850.9464
2697,22638.480,6,0.5,-1811.0784


In [75]:
df_clean.isnull().sum()

,0
Ship Mode,0
Segment,0
Country,0
City,0
State,0
Postal Code,0
Region,0
Category,0
Sub-Category,0
Sales,0


In [76]:
print("Remaining duplicates:", df_clean.duplicated().sum())

Remaining duplicates: 0


In [77]:
print("Final dataset shape:", df_clean.shape)

Final dataset shape: (9977, 13)


In [78]:
df_clean.dtypes

,0
Ship Mode,object
Segment,object
Country,object
City,object
State,object
Postal Code,int64
Region,object
Category,object
Sub-Category,object
Sales,float64


In [79]:
print("Sales minimum:", df_clean["Sales"].min())
print("Sales maximum:", df_clean["Sales"].max())

print("Quantity minimum:", df_clean["Quantity"].min())
print("Quantity maximum:", df_clean["Quantity"].max())

print("Discount minimum:", df_clean["Discount"].min())
print("Discount maximum:", df_clean["Discount"].max())

print("Profit minimum:", df_clean["Profit"].min())
print("Profit maximum:", df_clean["Profit"].max())

Sales minimum: 0.444
Sales maximum: 22638.48
Quantity minimum: 1
Quantity maximum: 14
Discount minimum: 0.0
Discount maximum: 0.8
Profit minimum: -6599.978
Profit maximum: 8399.976


In [80]:
print("Final shape:", df_clean.shape)
print("Total missing values:", df_clean.isnull().sum().sum())
print("Total duplicate rows:", df_clean.duplicated().sum())

Final shape: (9977, 13)
Total missing values: 0
Total duplicate rows: 0


In [81]:
df_clean.to_csv("cleaned_superstore.csv", index=False)

In [82]:
import os

print(os.path.exists("cleaned_superstore.csv"))

True


In [83]:
from google.colab import files

files.download("cleaned_superstore.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

#Data-cleaning conclusion.
17 exact duplicate rows were identified and removed. No missing values were present. Text fields were checked for whitespace and consistency, with no corrections required. Data types were appropriate for their respective columns. Quantity and Discount values were within valid ranges. Negative and zero Profit values were retained because they represent legitimate business outcomes. Statistical outliers in Sales and Profit were investigated and retained because there was no evidence that they were erroneous records

In [84]:
print("Final shape:", df_clean.shape)
print("Total missing values:", df_clean.isnull().sum().sum())
print("Total duplicate rows:", df_clean.duplicated().sum())

Final shape: (9977, 13)
Total missing values: 0
Total duplicate rows: 0


In [85]:
df_clean.to_csv("cleaned_superstore.csv", index=False)

In [86]:
import os

print(os.path.exists("cleaned_superstore.csv"))

True


In [87]:
from google.colab import files

files.download("cleaned_superstore.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# Data Cleaning Change Log

## Dataset
Superstore Dataset

## Cleaning Summary

| Issue Checked | Finding | Action Taken |
|---|---|---|
| Missing values | 0 missing values | No action required |
| Duplicate rows | 17 duplicate rows | Removed 17 duplicates |
| Text whitespace | No leading/trailing whitespace issues | No action required |
| Text consistency | Categories were consistently formatted | No action required |
| Data types | Appropriate data types | No conversion required |
| Quantity | Values ranged from 1 to 14 | No action required |
| Discount | Values ranged from 0.0 to 0.8 | No action required |
| Sales | Values ranged from 0.444 to 22,638.48 | No invalid negative values |
| Profit | Values ranged from -6,599.978 to 8,399.976 | Negative/zero values retained as legitimate business outcomes |
| Outliers | Potential statistical outliers identified | Investigated and retained because no data errors were identified |

## Final Dataset

- Original rows: 9,994
- Duplicate rows removed: 17
- Final rows: 9,977
- Total columns: 13
- Missing values: 0
- Duplicate rows remaining: 0